sonic-xcvrd/xcvrd/sff_mgr.py

```diff
@@ -0,0 +1,407 @@
+
+    def log_error(self, message):
+        self.helper_logger.log_error("SFF: {}".format(message))
+
+    def on_port_update_event(self, port_change_event):
+        if (port_change_event.event_type
+                not in [port_change_event.PORT_SET, port_change_event.PORT_DEL]):
+            return
+
+        lport = port_change_event.port_name
+        pport = port_change_event.port_index
+        asic_id = port_change_event.asic_id
+
+        # Skip if it's not a physical port
+        if not lport.startswith('Ethernet'):
+            return
+
+        # Skip if the physical index is not available
+        if pport is None:
+            return
+
+        if port_change_event.port_dict is None:
+            return
+
+        if port_change_event.event_type == port_change_event.PORT_SET:
+            if lport not in self.port_dict:
+                self.port_dict[lport] = {}
+            if pport >= 0:
+                self.port_dict[lport]['index'] = pport
+            # This field comes from CONFIG_DB PORT_TABLE. This is the channel
+            # that blongs to this logical port, 0 means all channels. tx_disable
+            # API needs to know which channels to disable/enable for a
+            # particular physical port.
+            if 'channel' in port_change_event.port_dict:
+                self.port_dict[lport]['channel'] = port_change_event.port_dict['channel']
+            # This field comes from STATE_DB PORT_TABLE
+            if 'host_tx_ready' in port_change_event.port_dict:
+                self.port_dict[lport]['host_tx_ready'] = \
+                        port_change_event.port_dict['host_tx_ready']
+            # This field comes from CONFIG_DB PORT_TABLE
+            if 'admin_status' in port_change_event.port_dict and \
+                port_change_event.db_name and \
+                port_change_event.db_name == 'CONFIG_DB':
+                # Only consider admin_status from CONFIG_DB.
+                # Ignore admin_status from STATE_DB, which may have
+                # different value.
+                self.port_dict[lport]['admin_status'] = \
+                        port_change_event.port_dict['admin_status']
+            # This field comes from STATE_DB TRANSCEIVER_INFO table.
+            # TRANSCEIVER_INFO has the same life cycle as a transceiver, if
+            # transceiver is inserted/removed, TRANSCEIVER_INFO is also
+            # created/deleted. Thus this filed can used to determine
+            # insertion/removal event.
+            if 'type' in port_change_event.port_dict:
+                self.port_dict[lport]['type'] = port_change_event.port_dict['type']
+            self.port_dict[lport]['asic_id'] = asic_id
+        # CONFIG_DB PORT_TABLE DEL case:
+        elif port_change_event.db_name and \
+                port_change_event.db_name == 'CONFIG_DB':
+            # Only when port is removed from CONFIG, we consider this entry as deleted.
+            if lport in self.port_dict:
```

Comment:
do you need do tx disable here in case the module is still present? According the HLD, tx disable would save power in such case.